include/Spectra/Util/RitzPairs.h

```diff
@@ -0,0 +1,110 @@
+//
+// This Source Code Form is subject to the terms of the Mozilla
+// Public License v. 2.0. If a copy of the MPL was not distributed
+// with this file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#ifndef SPECTRA_RITZ_PAIR_H
+#define SPECTRA_RITZ_PAIR_H
+
+#include <Eigen/Dense>
+
+namespace Spectra {
+template <typename Scalar>
+class SearchSpace;
+
+template <typename Scalar>
+class RitzPairs
+{
+private:
+    using Index = Eigen::Index;
+    using Matrix = Eigen::Matrix<Scalar, Eigen::Dynamic, Eigen::Dynamic>;
+    using Vector = Eigen::Matrix<Scalar, Eigen::Dynamic, 1>;
+    using Array = Eigen::Array<Scalar, Eigen::Dynamic, 1>;
+    using BoolArray = Eigen::Array<bool, Eigen::Dynamic, 1>;
+    using MapConstVec = Eigen::Map<const Vector>;
+
+public:
+    RitzPairs() = default;
+
+    void compute_eigen_pairs(SearchSpace<Scalar> const& search_space);
+
+    Array res_norm() const
+    {
+        return residues_.colwise().norm();
+    }  // norm of the residues
+
+    Index size() const { return values_.size(); }
+
+    void sort(SortRule selection)
+    {
+        std::vector<Index> ind = argsort(selection, values_);
+        RitzPairs<Scalar> temp = *this;
+        for (Index i = 0; i < size(); i++)
+        {
+            values_[i] = temp.values_[ind[i]];
+            vectors_.col(i) = temp.vectors_.col(ind[i]);
+            residues_.col(i) = temp.residues_.col(ind[i]);
+        }
+    }
+
+    bool check_convergence(Scalar tol, Index number_eigenvalues) const
+    {
+        const Array norms = res_norm();
+        bool converged = true;
+        root_converged_ = BoolArray::Zero(norms.size());
+        for (Index j = 0; j < norms.size(); j++)
+        {
+            root_converged_[j] = (norms[j] < tol);
+            if (j < number_eigenvalues)
+            {
+                converged &= (norms[j] < tol);
+            }
+        }
+        return converged;
+    }
+
+    const Matrix& RitzVectors() const { return vectors_; }
+    const Vector& RitzValues() const { return values_; }
+    const Matrix& SmallRitzVectors() const { return small_vectors_; }
+    const Matrix& Residues() const { return residues_; }
+    const BoolArray& ConvergedEigenvalues() const { return root_converged_; }
+
+private:
+    Vector values_;         // eigenvalues
+    Matrix small_vectors_;  // eigenvectors of the small problem, makes restart cheaper.
+    Matrix vectors_;        // Ritz (or harmonic Ritz) eigenvectors
+    Matrix residues_;       // residues of the pairs
+    BoolArray root_converged_;
+};
+
+}  // namespace Spectra
+#include "SearchSpace.h"
+namespace Spectra {
+
+template <typename Scalar>
+void RitzPairs<Scalar>::compute_eigen_pairs(SearchSpace<Scalar> const& search_space)
+{
+    const Matrix& basis_vectors = search_space.BasisVectors();
+    const Matrix& op_basis_prod = search_space.OperatorBasisProduct();
+
+    // form the small eigenvalue
+    Matrix small_matrix = basis_vectors.transpose() * op_basis_prod;
```

Comment:
i still think we should store and update the small space matrix ... If the operator size is 10^6 and we have 100 basis vector re-computing the V^T A V projection is expensive

Reply:
Plus it's only one new small member variable in the RitzPair class

Reply:
fair point, you or me?

Reply:
Can you do it ? I have a meeting starting in 20 min

Reply:
okay :)

Reply:
so we push that back, because it would require knowledge about the exact orthogonalisation method.

Reply:
that being said it doesn't since the two updates functions : `update_operator_basis_product` and `full_update` are respectively for QR and GS. So in `update_operator_basis_product` we could implement the update of the small matrix while in `full_update` we will simply recompute the full projection.